Ruleset methods

elegaanz opened the pull request:
The API is quite different from what was proposed in the issue, since it was not possible to determine in which field of `Ruleset` to put a new `AnyPushRule`. I had to create wrapper types for each kind of push rules, to be able to tell at compile-time where in the rule set to store a push rule.

Fixes #352

elegaanz committed:
Use a BTreeSet instead of a Vec to make push rule IDs unique

- Introduce a new macro to easily derive Ord, PartialOrd, Eq and PartialEq
  for a PushRule-like type (these traits are required by BTreeSet)
- Introduce a new macro to easily build a BTreeSet

elegaanz committed:
Implement IntoIterator for Ruleset

elegaanz committed:
Add a "add" method on Ruleset

- Add wrapper types for each kind of push rule
- Add a trait to add a push rule to a rule set

elegaanz committed:
impl Extend for Ruleset

jplatte commented on ruma-common/src/push/predefined.rs:
```diff
@@ -8,6 +8,21 @@ use super::{
 
 use ruma_identifiers::UserId;
 
+macro_rules! set {
```
We're already using the `maplit` crate in other Ruma crates, might as well add a dependency here too and use `btreeset` from there.

jplatte commented on ruma-common/src/push.rs:
```diff
@@ -81,6 +81,20 @@ impl Ruleset {
+            .chain(self.room.into_iter().map(Into::into))
+            .chain(self.sender.into_iter().map(Into::into))
+            .chain(self.underride.into_iter().map(Into::into))
+            .collect::<Vec<_>>().into_iter()
```
If we put a bit more work into this, it can work without allocating. But we can also defer that to later.

elegaanz replied:
I will try something.

jplatte commented on ruma-common/src/push.rs:
```diff
@@ -1,6 +1,38 @@
+//! - sender rules
+//!
+//! Each of these kind of rule has a corresponding type, that is
+//! just a wrapper arround another type:
```
Extraneous comma, extra 'r' in `around`.

jplatte commented on ruma-common/src/push/predefined.rs:
```diff
@@ -34,45 +34,45 @@ impl Ruleset {
     /// notifications other than those generated by override rules set by the user.
     pub fn master() -> Self {
-        Self {
+        OverridePushRule(ConditionalPushRule {
```
```suggestion
        Self(ConditionalPushRule {
```

elegaanz replied:
I tried that, but apparently it is not a valid syntax. :woman_shrugging:

elegaanz replied:
Ok, I just re-tried and it works, my bad.

jplatte reviewed (commented):
This looks super awesome, thanks!

I was actually imagining `add` taking an `AnyPushRule` (with that then having five variants rather than three). Then we can also have `impl Extend<AnyPushRule>` for Ruleset` (and the other `Extend` implementations can be implemented in terms of that rather than requiring an extra trait).

One more thing: Can you rename `PushRule` to `SimplePushRule` or `BasicPushRule` while you're at it?

elegaanz commented:
> I was actually imagining `add` taking an `AnyPushRule` (with that then having five variants rather than three). Then we can also have `impl Extend<AnyPushRule>` for Ruleset`(and the other`Extend` implementations can be implemented in terms of that rather than requiring an extra trait).

AnyPushRule is not an enum, it is a struct, that's why I didn't added variants here.

> One more thing: Can you rename `PushRule` to `SimplePushRule` or `BasicPushRule` while you're at it?

Yes, I was thinking of doing that too, but I wasn't sure if it was a really a good idea or not. I will do it.

elegaanz committed:
Rename PushRule to SimplePushRule

jplatte commented:
> AnyPushRule is not an enum, it is a struct, that's why I didn't added variants here.

Right, and making it one isn't straight-forward because we wouldn't be able to tell which variant to choose in deserialization. There may be ways around that, but that's out of scope for this PR. In that case, can you [seal](https://rust-lang.github.io/api-guidelines/future-proofing.html) the trait you added?

elegaanz committed:
Use Self constructors when possible

elegaanz committed:
Seal the RulesetMember trait

+ use maplit instead of custom macro
+ fix two little issues in the documentation

elegaanz committed:
Run cargo fmt

jplatte commented on ruma-common/src/push.rs:
```diff
@@ -94,27 +94,62 @@ impl Ruleset {
-pub trait RulesetMember {
+pub trait RulesetMember: private::Sealed {
     /// Adds a value in the correct field of a Ruleset.
     fn add_to(self, ruleset: &mut Ruleset) -> bool;
```
Can you also add `#[doc(hidden)]` to this function, since users aren't meant to call it directly?

jplatte commented:
PR needs to be rebased, but other than that and hiding the `add_to` function, this looks ready to merge :+1:

elegaanz committed:
Hide RulesetMember::add_to in the docs

elegaanz committed:
Custom iterator for Ruleset

elegaanz commented:
Rebased and added the `#[doc(hidden)]`
